File: packages/b2c-tooling-sdk/src/operations/jobs/run-system-job.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * Dual-backend runner for B2C Commerce **system jobs** (site-archive
 * import/export, CAP install/uninstall).
 *
 * These operations all follow the same shape: trigger a named system job,
 * wait for it to finish, and surface the job log on failure. This module
 * centralizes that flow over either backend so each operation declares only
 * its job ID and request body — not the OCAPI-vs-SCAPI plumbing.
 *
 * ## Contract
 *
 * The public result/error types of every caller expose the **raw OCAPI**
 * {@link JobExecution} (snake_case). To keep that contract identical across
 * backends, the SCAPI path maps its canonical result back to the OCAPI shape
 * via {@link mapCanonicalToOcapiExecution}, and a SCAPI job failure is
 * re-thrown as the raw {@link JobExecutionError} — so a caller's existing
 * `catch (JobExecutionError) → getJobLog(instance, err.execution)` works
 * unchanged regardless of which backend served the job.
 *
 * ## Backend selection & fallback
 *
 * Honors {@link B2CInstance.apiBackend}:
 *   - `'ocapi'`: always OCAPI.
 *   - `'scapi'`: always SCAPI (throws if the instance can't reach SCAPI).
 *   - `'auto'` (default): SCAPI when {@link B2CInstance.scapiClientConfig} is
 *     available, else OCAPI. **Fallback to OCAPI happens only if the SCAPI
 *     *start* fails** (missing scope, body rejected, system job not
 *     triggerable over SCAPI). Once a job has started, the wait/log path never
 *     falls back — re-running a started write would be unsafe.
 *
 * ## Lifecycle
 *
 * Lives alongside the other transitional jobs plumbing. When OCAPI is removed,
 * delete the OCAPI branch and inline the SCAPI calls.
 *
 * @module operations/jobs/run-system-job
 */
import type {B2CInstance, ScapiClientConfig} from '../../instance/index.js';
import {isOcapiDeprecatedFault, OcapiDeprecatedError} from '../../clients/error-utils.js';
import {createScapiJobsClient} from '../../clients/scapi-jobs.js';
import {getLogger} from '../../logging/logger.js';
import {mapCanonicalToOcapiExecution} from './ocapi-mapping.js';
import {executeJob as scapiExecuteJob, getJobExecution as scapiGetJobExecution} from './scapi-ops.js';
import {waitForJob, JobExecutionError, type JobExecution, type WaitForJobOptions} from './run.js';
import {waitForJobExecution, CanonicalJobExecutionError} from './wait-canonical.js';

/**
 * Declarative description of a system job to run. The operation supplies its
 * job ID and the two request-body forms; this module drives the rest.
 */
export interface SystemJobSpec {
  /** System job ID, e.g. `sfcc-site-archive-import`. */
  jobId: string;
  /**
   * OCAPI "shorthand" request body tried first on the OCAPI path (e.g.
   * `{file_name}`, `{export_file, data_units}`, or the CAP `{app_name, ...}`
   * shape). When the instance rejects it with `UnknownPropertyException`, the
   * OCAPI path retries with {@link parameters}.
   */
  ocapiBody: Record<string, unknown>;
  /**
   * Job parameters (`[{name, value}]`). Used as the OCAPI internal-user retry
   * body **and** as the SCAPI request body (SCAPI's `JobExecutionRequest`
   * accepts exactly this shape).
   */
  parameters: Array<{name: string; value: string}>;
  /** Scopes named in the OCAPI-deprecation error (the rw jobs scope). */
  deprecatedScopes?: string[];
  /** Whether to wait for completion (default `true`). */
  wait?: boolean;
  /** Wait options forwarded to the poll loop. */
  waitOptions?: WaitForJobOptions;
  /** Human-readable verb for error messages, e.g. `'execute import job'`. */
  failVerb: string;
}

/**
 * Starts a system job (waiting for completion unless `spec.wait === false`)
 * and returns the raw OCAPI {@link JobExecution}. Throws {@link JobExecutionError}
 * (raw) when the job fails, or {@link OcapiDeprecatedError} when the only
 * reachable backend is a deprecated OCAPI.
 */
export async function runSystemJob(instance: B2CInstance, spec: SystemJobSpec): Promise<JobExecution> {
  const preference = instance.apiBackend;
  const scapiConfig = instance.scapiClientConfig;

  if (preference === 'ocapi') {
    return runOcapiSystemJob(instance, spec);
  }

  if (preference === 'scapi') {
    if (!scapiConfig) {
      throw new Error(
        `${spec.jobId} SCAPI backend requires shortCode, tenantId, and OAuth credentials. ` +
          `Configure them in dw.json or set apiBackend to ocapi.`,
      );
    }
    return runScapiSystemJob(instance, scapiConfig, spec);
  }

  // auto
  if (!scapiConfig) {
    return runOcapiSystemJob(instance, spec);
  }

  // Try SCAPI start; fall back to OCAPI only if the start fails (nothing has
  // run yet). Once started, finishScapiJob handles wait/failure without
  // falling back.
  const client = createScapiJobsClient(
    {shortCode: scapiConfig.shortCode, tenantId: scapiConfig.tenantId},
    scapiConfig.auth,
  );
  let started;
  try {
    started = await startScapiJob(client, scapiConfig.tenantId, spec);
  } catch (error) {
    getLogger().info(
      {jobId: spec.jobId, reason: error instanceof Error ? error.message : String(error)},
      `SCAPI ${spec.jobId} unavailable, falling back to OCAPI`,
    );
    return runOcapiSystemJob(instance, spec);
  }
  return finishScapiJob(client, scapiConfig.tenantId, spec, started);
}

/**
 * SCAPI path: trigger the job (start phase, fallback-eligible in auto) then
 * wait/map (finish phase, never falls back).
 */
async function runScapiSystemJob(
  instance: B2CInstance,
  scapiConfig: ScapiClientConfig,
  spec: SystemJobSpec,
): Promise<JobExecution> {
  const client = createScapiJobsClient(
    {shortCode: scapiConfig.shortCode, tenantId: scapiConfig.tenantId},
    scapiConfig.auth,
  );
  const started = await startScapiJob(client, scapiConfig.tenantId, spec);
  return finishScapiJob(client, scapiConfig.tenantId, spec, started);
}

async function startScapiJob(client: ReturnType<typeof createScapiJobsClient>, tenantId: string, spec: SystemJobSpec) {
  getLogger().debug({jobId: spec.jobId}, `Executing ${spec.jobId} job via SCAPI`);
  return scapiExecuteJob(client, spec.jobId, {parameters: spec.parameters, tenantId});
}

async function finishScapiJob(
  client: ReturnType<typeof createScapiJobsClient>,
  tenantId: string,
  spec: SystemJobSpec,
  started: Awaited<ReturnType<typeof scapiExecuteJob>>,
): Promise<JobExecution> {
  getLogger().debug({jobId: spec.jobId, executionId: started.id}, `${spec.jobId} job started: ${started.id}`);

  if (spec.wait === false) {
    return mapCanonicalToOcapiExecution(started);
  }

  try {
    const final = await waitForJobExecution(
      (jobId, executionId) => scapiGetJobExecution(client, jobId, executionId, tenantId),
      spec.jobId,
      started.id,
      spec.waitOptions,
    );
    return mapCanonicalToOcapiExecution(final);
  } catch (error) {
    // Re-throw a job FAILURE as the raw JobExecutionError so callers' existing
    // log-fetch handling works identically across backends. Other errors
    // (timeout, network) propagate as-is — never fall back post-start.
    if (error instanceof CanonicalJobExecutionError) {
      throw new JobExecutionError(error.message, mapCanonicalToOcapiExecution(error.execution));
    }
    throw error;
  }
}

/**
 * OCAPI path: preserves the legacy behavior exactly — POST the shorthand body,
 * retry with the parameters body on `UnknownPropertyException`, then wait.
 */
async function runOcapiSystemJob(instance: B2CInstance, spec: SystemJobSpec): Promise<JobExecution> {
  const logger = getLogger();
  logger.debug({jobId: spec.jobId}, `Executing ${spec.jobId} job via OCAPI`);

  let execution: JobExecution;

  const {data, error} = await instance.ocapi.POST('/jobs/{job_id}/executions', {
    params: {path: {job_id: spec.jobId}},
    body: spec.ocapiBody as unknown as string,
  });

  if (
    error?.fault?.type === 'UnknownPropertyException' &&
    (error.fault.arguments as Record<string, unknown>)?.document === 'job_execution_request'
  ) {
    // Retry with parameters format (internal/support users)
    logger.warn('Retrying with parameters format for internal users');

    const {data: retryData, error: retryError} = await instance.ocapi.POST('/jobs/{job_id}/executions', {
      params: {path: {job_id: spec.jobId}},
      body: {parameters: spec.parameters} as unknown as string,
    });

    if (retryError || !retryData) {
      if (isOcapiDeprecatedFault(retryError))
        throw new OcapiDeprecatedError({cause: retryError, requiredScopes: spec.deprecatedScopes});
      throw new Error(retryError?.fault?.message ?? `Failed to ${spec.failVerb}`, {cause: retryError});
    }

    execution = retryData;
  } else if (error || !data) {
    if (isOcapiDeprecatedFault(error))
      throw new OcapiDeprecatedError({cause: error, requiredScopes: spec.deprecatedScopes});
    throw new Error(error?.fault?.message ?? `Failed to ${spec.failVerb}`, {cause: error});
  } else {
    execution = data;
  }

  logger.debug({jobId: spec.jobId, executionId: execution.id}, `${spec.jobId} job started: ${execution.id}`);

  if (spec.wait === false) {
    return execution;
  }

  return waitForJob(instance, spec.jobId, execution.id!, spec.waitOptions);
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/operations/jobs/run_system_job.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Dual-backend runner for B2C Commerce system jobs.

Mirrors ``src/operations/jobs/run-system-job.ts``. System jobs (site-archive
import/export, CAP install/uninstall) all follow the same shape: trigger a named
system job, wait for it to finish, and surface the job log on failure. This
module centralizes that flow over either backend so each operation declares only
its job ID and request body — not the OCAPI-vs-SCAPI plumbing.

The public result/error types of every caller expose the raw OCAPI
:data:`JobExecution` (snake_case). The SCAPI path maps its canonical result back
to the OCAPI shape via :func:`map_canonical_to_ocapi_execution`, and a SCAPI job
failure is re-raised as the raw :class:`JobExecutionError`.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from b2c_tooling_sdk.clients.error_utils import OcapiDeprecatedError, is_ocapi_deprecated_fault
from b2c_tooling_sdk.clients.scapi_backend_utils import is_fallback_trigger, scapi_unavailable_message
from b2c_tooling_sdk.clients.scapi_jobs import ScapiJobsClient, ScapiJobsClientConfig, create_scapi_jobs_client
from b2c_tooling_sdk.logging import get_logger
from b2c_tooling_sdk.operations.jobs.ocapi_mapping import map_canonical_to_ocapi_execution
from b2c_tooling_sdk.operations.jobs.run import (
    JobExecution,
    JobExecutionError,
    WaitForJobOptions,
    wait_for_job,
)
from b2c_tooling_sdk.operations.jobs.scapi_ops import ExecuteJobScapiOptions
from b2c_tooling_sdk.operations.jobs.scapi_ops import execute_job as scapi_execute_job
from b2c_tooling_sdk.operations.jobs.scapi_ops import get_job_execution as scapi_get_job_execution
from b2c_tooling_sdk.operations.jobs.types import JobExecutionInfo
from b2c_tooling_sdk.operations.jobs.wait_canonical import CanonicalJobExecutionError, wait_for_job_execution

if TYPE_CHECKING:
    from b2c_tooling_sdk.instance import B2CInstance, ScapiClientConfig


def _is_safe_start_fallback(error: object) -> bool:
    """Return ``True`` when a SCAPI start failure provably created no job."""
    return is_fallback_trigger(error)


@dataclass
class SystemJobSpec:
    """Declarative description of a system job to run."""

    #: System job ID, e.g. ``sfcc-site-archive-import``.
    job_id: str
    #: OCAPI "shorthand" request body tried first on the OCAPI path. When the
    #: instance rejects it with ``UnknownPropertyException``, the OCAPI path
    #: retries with :attr:`parameters`.
    ocapi_body: dict[str, Any]
    #: Job parameters (``[{name, value}]``). Used as the OCAPI internal-user
    #: retry body and as the SCAPI request body.
    parameters: list[dict[str, str]]
    #: Human-readable verb for error messages, e.g. ``"execute import job"``.
    fail_verb: str
    #: Scopes named in the OCAPI-deprecation error (the rw jobs scope).
    deprecated_scopes: list[str] | None = None
    #: Whether to wait for completion (default: ``True``).
    wait: bool = True
    #: Wait options forwarded to the poll loop.
    wait_options: WaitForJobOptions | None = None


async def run_system_job(instance: B2CInstance, spec: SystemJobSpec) -> JobExecution:
    """Start a system job (waiting for completion unless ``spec.wait`` is ``False``).

    :returns: the raw OCAPI :data:`JobExecution`.
    :raises JobExecutionError: (raw) when the job fails.
    :raises OcapiDeprecatedError: when the only reachable backend is a deprecated OCAPI.
    """
    preference = instance.api_backend
    scapi_config = instance.scapi_client_config

    if preference == "ocapi":
        return await _run_ocapi_system_job(instance, spec)

    if preference == "scapi":
        if not scapi_config:
            # Domain label (not the job ID) so the message reads "Jobs SCAPI
            # backend requires...", consistent with resolve_scapi_or_ocapi.
            raise RuntimeError(scapi_unavailable_message("Jobs"))
        return await _run_scapi_system_job(instance, scapi_config, spec)

    # auto
    if not scapi_config:
        return await _run_ocapi_system_job(instance, spec)

    client = create_scapi_jobs_client(
        ScapiJobsClientConfig(short_code=scapi_config.short_code, tenant_id=scapi_config.tenant_id),
        scapi_config.auth,
    )
    try:
        started = await _start_scapi_job(client, scapi_config.tenant_id, spec)
    except Exception as error:  # noqa: BLE001 - classified below
        # Fall back to OCAPI ONLY when the SCAPI start provably created no job.
        # An ambiguous failure (network drop after dispatch, timeout, 5xx) must
        # propagate — re-running a mutating system job over OCAPI could execute
        # it twice.
        if not _is_safe_start_fallback(error):
            raise
        get_logger("operations.jobs").info("SCAPI %s start rejected, falling back to OCAPI: %s", spec.job_id, error)
        return await _run_ocapi_system_job(instance, spec)
    return await _finish_scapi_job(client, scapi_config.tenant_id, spec, started)


async def _run_scapi_system_job(
    instance: B2CInstance,
    scapi_config: ScapiClientConfig,
    spec: SystemJobSpec,
) -> JobExecution:
    client = create_scapi_jobs_client(
        ScapiJobsClientConfig(short_code=scapi_config.short_code, tenant_id=scapi_config.tenant_id),
        scapi_config.auth,
    )
    started = await _start_scapi_job(client, scapi_config.tenant_id, spec)
    return await _finish_scapi_job(client, scapi_config.tenant_id, spec, started)


async def _start_scapi_job(client: ScapiJobsClient, tenant_id: str, spec: SystemJobSpec) -> JobExecutionInfo:
    return await scapi_execute_job(
        client,
        spec.job_id,
        ExecuteJobScapiOptions(tenant_id=tenant_id, parameters=spec.parameters),
    )


async def _finish_scapi_job(
    client: ScapiJobsClient,
    tenant_id: str,
    spec: SystemJobSpec,
    started: JobExecutionInfo,
) -> JobExecution:
    if spec.wait is False:
        return map_canonical_to_ocapi_execution(started)

    async def _get(job_id: str, execution_id: str) -> JobExecutionInfo:
        return await scapi_get_job_execution(client, job_id, execution_id, tenant_id)

    try:
        final = await wait_for_job_execution(
            _get,
            spec.job_id,
            started.id,
            spec.wait_options or WaitForJobOptions(),
        )
        return map_canonical_to_ocapi_execution(final)
    except CanonicalJobExecutionError as error:
        # Re-raise a job FAILURE as the raw JobExecutionError so callers' existing
        # log-fetch handling works identically across backends. Other errors
        # (timeout, network) propagate as-is — never fall back post-start.
        raise JobExecutionError(str(error), map_canonical_to_ocapi_execution(error.execution)) from error


async def _run_ocapi_system_job(instance: B2CInstance, spec: SystemJobSpec) -> JobExecution:
    execution: JobExecution

    result = await instance.ocapi.post(
        "/jobs/{job_id}/executions",
        {"params": {"path": {"job_id": spec.job_id}}, "body": spec.ocapi_body},
    )
    data, error = result.data, result.error

    fault = error.get("fault") if isinstance(error, dict) else None
    arguments = fault.get("arguments") if isinstance(fault, dict) else None
    is_unknown_property = (
        isinstance(fault, dict)
        and fault.get("type") == "UnknownPropertyException"
        and isinstance(arguments, dict)
        and arguments.get("document") == "job_execution_request"
    )

    if is_unknown_property:
        # Retry with parameters format (internal/support users).
        retry = await instance.ocapi.post(
            "/jobs/{job_id}/executions",
            {"params": {"path": {"job_id": spec.job_id}}, "body": {"parameters": spec.parameters}},
        )
        retry_data, retry_error = retry.data, retry.error
        if retry_error or not retry_data:
            if is_ocapi_deprecated_fault(retry_error):
                raise OcapiDeprecatedError(cause=retry_error, required_scopes=spec.deprecated_scopes)
            raise RuntimeError(_fault_message(retry_error) or f"Failed to {spec.fail_verb}")
        execution = retry_data
    elif error or not data:
        if is_ocapi_deprecated_fault(error):
            raise OcapiDeprecatedError(cause=error, required_scopes=spec.deprecated_scopes)
        raise RuntimeError(_fault_message(error) or f"Failed to {spec.fail_verb}")
    else:
        execution = data

    if spec.wait is False:
        return execution

    return await wait_for_job(instance, spec.job_id, execution["id"], spec.wait_options or WaitForJobOptions())


def _fault_message(error: Any) -> str | None:
    if isinstance(error, dict):
        fault = error.get("fault")
        if isinstance(fault, dict):
            message = fault.get("message")
            if isinstance(message, str) and message:
                return message
    return None


__all__ = ["SystemJobSpec", "run_system_job"]
